### PR TITLE
[WIP] Check nonempty nulls in output of `set_null_mask`

### PR DESCRIPTION
**Pull request by cindyyuanjiang**
## Description
We check for nonempty nulls in the output of `set_null_mask` and call `purge_nonempty_nulls` to clean up the column.
Fixes https://github.com/NVIDIA/spark-rapids/issues/7485.

We currently investigated the performance impact of adding this check by comparing the runtime of multiple integration tests in the Spark plugin side which calls `mergeAndSetValidity`, and the factor of runtime after adding this check over the runtime before ranges from 1.007 to 1.032.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: java/src/main/native/src/ColumnViewJni.cpp [hunk 1]
```diff
@@ -1361,6 +1361,10 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnView_binaryOpVV(JNIEnv *env, j
         out->set_null_mask(std::move(new_mask), null_count);
       }
 
+      if (cudf::may_have_nonempty_nulls(*out)) {
```
Comment (revans2) [inline java/src/main/native/src/ColumnViewJni.cpp:1364]: This is not needed here. Only fixed width types are supported for `out` as can be seen on line 1355, and fixed width types cannot have a nonempty null.

File: java/src/main/native/src/ColumnViewJni.cpp [hunk 2]
```diff
@@ -1407,6 +1411,10 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnView_binaryOpVS(JNIEnv *env, j
         out->set_null_mask(std::move(new_mask), new_null_count);
       }
 
+      if (cudf::may_have_nonempty_nulls(*out)) {
```
Comment (revans2) [inline java/src/main/native/src/ColumnViewJni.cpp:1414]: Same comment here. This is not needed because it is only working on fixed width types.

Comment (cindyyuanjiang) [reply java/src/main/native/src/ColumnViewJni.cpp]: Thank you! Changes updated.

File: java/src/main/native/src/ColumnViewJni.cpp [hunk 3]
```diff
@@ -1786,6 +1786,10 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnView_bitwiseMergeAndSetValidit
       default: JNI_THROW_NEW(env, cudf::jni::ILLEGAL_ARG_CLASS, "Unsupported merge operation", 0);
     }
 
+    if (cudf::may_have_nonempty_nulls(*copy)) {
```
Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cpp:1789]: `may_have_nonempty_nulls` is more conservative and still causes performance regression. I've tried it and failed. Please use `cudf::has_nonempty_nulls`.

**Comment by cindyyuanjiang** [comment]
Here are some performance benchmark results comparing this PR (version with `may_have_nonempty_nulls`) against branch-23.04:
--------------------------------------------------------------------
Name = query14_part1
Means = 30578.342723846436, 31357.60169029236
Time diff = -779.2589664459229
Speedup = 0.975149280415563
T-Test (test statistic, p value, df) = -3.6117985371970183, 0.006863742633385148, 8.0
T-Test Confidence Interval = -1276.7879344599287, -281.72999843191695
ALERT: significant change has been detected (p-value < 0.05)
ALERT: regression in performance has been observed
--------------------------------------------------------------------
Name = query23_part1
Means = 43601.79753303528, 42367.023849487305
Time diff = 1234.7736835479736
Speedup = 1.0291446878103738
T-Test (test statistic, p value, df) = 2.3231823002195817, 0.04867700901728152, 8.0
T-Test Confidence Interval = 9.13021173653442, 2460.417155359413
ALERT: significant change has been detected (p-value < 0.05)
ALERT: improvement in performance has been observed

File: java/src/main/native/src/ColumnViewJni.cpp (hunk 3 above)
Comment (cindyyuanjiang) [reply java/src/main/native/src/ColumnViewJni.cpp]: Updated. Thank you!

File: java/src/main/native/src/ColumnViewJni.cpp [hunk 4]
```diff
@@ -1833,6 +1833,10 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnView_bitwiseMergeAndSetValidit
       default: JNI_THROW_NEW(env, cudf::jni::ILLEGAL_ARG_CLASS, "Unsupported merge operation", 0);
     }
 
+    if (cudf::has_nonempty_nulls(*copy)) {
+      copy = cudf::purge_nonempty_nulls(*copy);
+    }
```
Comment (ttnghia) [inline java/src/main/native/src/ColumnViewJni.cpp:1838]: Avoid deref `copy` twice: that will create two instances of `column_view`.
```suggestion
    if (auto const copy_cv = copy->view(); cudf::has_nonempty_nulls(copy_cv)) {
      copy = cudf::purge_nonempty_nulls(copy_cv);
    }
```

**Comment by razajafri** [comment]
closed in favor of #13335

**Pull request closed without merging**